File: problems/matmul/matmul.cpp
#include <stdio.h>
#include <stdlib.h>
#include <vector>

void matmul(int n, const std::vector<double> &a, const std::vector<double> &b,
            std::vector<double> &c) {
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
      double sum = 0.0;
      for (int k = 0; k < n; k++) {
        sum += a[i * n + k] * b[k * n + j];
      }
      c[i * n + j] = sum;
    }
  }
}

int main(int argc, char **argv) {
  int n = 512;
  std::vector<double> a(n * n, 1.0 * argc);
  std::vector<double> b(n * n, 2.0);
  std::vector<double> c(n * n);

  matmul(n, a, b, c);

  double checksum = 0.0;
  for (double val : c) {
    checksum += val;
  }

  if (checksum != (268435456.0 * argc)) {
    return 1;
  }

  return 0;
}

File: problems/matmul/matmul.rs
fn matmul(n: usize, a: &[f64], b: &[f64], c: &mut [f64]) {
    for i in 0..n {
        for k in 0..n {
            for j in 0..n {
                c[i * n + j] += a[i * n + k] * b[k * n + j];
            }
        }
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let argc = args.len() as f64;

    let n = 1024;
    let a = vec![1.0 * argc; n * n];
    let b = vec![2.0; n * n];
    let mut c = vec![0.0; n * n];

    matmul(n, &a, &b, &mut c);
    
    let checksum: f64 = c.iter().sum();
    if checksum != (2147483648.0 * argc) {
        std::process::exit(1);
    }
}
